plugin-server/src/worker/ingestion/event-pipeline/extractHeatmapDataStep.ts

```diff
@@ -0,0 +1,122 @@
+import { URL } from 'url'
+
+import { PreIngestionEvent, RawClickhouseHeatmapEvent, TimestampFormat } from '../../../types'
+import { castTimestampOrNow } from '../../../utils/utils'
+import { EventPipelineRunner } from './runner'
+
+// This represents the scale factor for the heatmap data. Essentially how much we are reducing the resolution by.
+const SCALE_FACTOR = 16
+
+type HeatmapDataItem = {
+    x: number
+    y: number
+    target_fixed: boolean
+    type: string
+}
+
+type HeatmapData = Record<string, HeatmapDataItem[]>
+
+export function extractHeatmapDataStep(
+    runner: EventPipelineRunner,
+    event: PreIngestionEvent
+): Promise<[PreIngestionEvent, Promise<void>[]]> {
+    const { eventUuid } = event
+
+    let acks: Promise<void>[] = []
+
+    try {
+        const heatmapEvents = extractScrollDepthHeatmapData(event) ?? []
+
+        // eslint-disable-next-line @typescript-eslint/no-floating-promises
+        acks = heatmapEvents.map((rawEvent) => {
+            return runner.hub.kafkaProducer.produce({
+                topic: runner.hub.CLICKHOUSE_HEATMAPS_KAFKA_TOPIC,
+                key: eventUuid,
+                value: Buffer.from(JSON.stringify(rawEvent)),
+                waitForAck: true,
+            })
+        })
+    } catch (e) {
+        console.error('Error extracting heatmap data:', e, event)
```

Comment:
This is to keep it "safe" in that we don't want to drop events due to bad heatmap parsing. Not sure what the standard would be here for plugin server? Sentry exception? Ingestion warning?

Reply:
I guess ingestion warning... since that's got built in rate limiting (you can specify additions to the limiter key to tune it)